### ISSUE TITLE
Model instance added to array after afterAttach() results in wrong parent assumptions and patches order

### ISSUE THREAD

**Issue by andrezzoid** [mobxjs/mobx-state-tree#1782]
**_Bug report_**

* [x] I've checked documentation and searched for existing issues
* [x] I've made sure my project is based on the latest MST version
* [x] Fork [this](https://codesandbox.io/s/y64pzxj01) code sandbox or another minimal reproduction. 

**Sandbox link or minimal reproduction code**

https://codesandbox.io/s/relaxed-zhukovsky-przqn?file=/src/todos.test.js

Open the terminal and run `npm test` to see both tests failing.

**Describe the expected behavior**

As per the the [documentation](https://mobx-state-tree.js.org/overview/hooks):
> So, in afterAttach one can safely make assumptions about the parent (...)

When adding a Model to an Array type, it would be expected for the Model instance to be added to the array before `afterAttach` is called so one can "safely make assumptions about the parent".

**Describe the observed behavior**

In the example sandbox we have a `TodoList` which has an array `todos` of `Todo`'s. When a `Todo`is being attached it cannot fully make assumptions about the parent array in `afterAttach` because it hasn't actually been attached yet. One wrong assumption would be to measure the parent's `todos.length`. The above referenced documentation is only half true here. 

As a side effect, in case there are changes to the `Todo` instance inside `afterAttach`, the recorded patches for `TodoList` will show these changes happening before the `Todo` is added to the array.

```
[
  { op: 'replace', path: '/todos/0/title', value: 'New Todo' },
  {
    op: 'add',
    path: '/todos/0',
    value: { title: 'New Todo', done: true }
  }
]
```

**Comment by coolsoftwaretyler** [comment]
Hey @andrezzoid - sorry it took so long for anyone to get back to you here. 

I think that https://github.com/mobxjs/mobx-state-tree/pull/1952 is a likely fix for this. So I'm going to label this as `has PR` for now. That PR is held up a little bit in maintenance. I'm not sure when it will get on the roadmap, but it's clearly a widespread issue that we will try and resolve. I think the big question is how breaking/not breaking that change will be.

If you've since moved on past this issue, that's totally understandable. But if you have some time, it would be awesome if you could see how that PR does or does not address your issue, and let us know!

**Comment by coolsoftwaretyler** [comment]
We merged #1952, so I'm gonna close this out. If you are still experiencing this issue, please let me know!
